x/lockup/keeper/lock.go

```diff
@@ -904,3 +904,76 @@ func (k Keeper) getCoinsFromLocks(locks []types.PeriodLock) sdk.Coins {
 	}
 	return coins
 }
+
+// RebondTokens rebonds some amount or all coins of a lock with id lockID only if it is already unbonding
+func (k Keeper) RebondTokens(ctx sdk.Context, lockID uint64, owner sdk.AccAddress, coins sdk.Coins) error {
+	lock, err := k.GetLockByID(ctx, lockID)
+	if err != nil {
+		return err
+	}
+
+	// check synthetic lockup exists
+	if k.HasAnySyntheticLockups(ctx, lock.ID) {
+		// TODO: export this error
+		return fmt.Errorf("cannot edit lockup with synthetic lock %d", lock.ID)
+	}
+
+	if lock.Owner != owner.String() {
+		return fmt.Errorf("lock %d is not owned by %s", lockID, owner)
+	}
+
+	if !lock.IsUnlocking() {
+		return errorsmod.Wrap(types.ErrLockNotUnlocking, fmt.Sprintf("lock %d is not unlocking, rebonding only possible in unlocking stage", lockID))
+	}
+
+	// If all checks pass, we can rebond the tokens
+	var rebondedLock types.PeriodLock
+	// partial re-bonding
+	if coins != nil && !coins.IsEqual(lock.Coins) {
+		// This branch implies we will end up with two locks:
+		// - The first lock will be the original lock with the coins removed, in unlocking state
+		// - The second lock will be the new rebonded lock with the coins added, in locked state
+
+		// remove rebonded lock's coins from original lock
+		lock.Coins = lock.Coins.Sub(coins)
+		err = k.setLock(ctx, *lock)
+		if err != nil {
+			return errorsmod.Wrap(types.ErrFailedToSetLock, fmt.Sprintf("failed to set original lock %d: %s", lock.ID, err.Error()))
+		}
+
+		// create a rebonded lock
+		splitLockID := k.GetLastLockID(ctx) + 1
+		k.SetLastLockID(ctx, splitLockID)
+		rebondedLock = types.NewPeriodLock(splitLockID, owner, owner.String(), lock.Duration, time.Time{}, coins)
+
+		err = k.setLock(ctx, rebondedLock)
+		if err != nil {
+			return errorsmod.Wrap(types.ErrFailedToSetLock, fmt.Sprintf("failed to set rebonding lock %d: %s", rebondedLock.ID, err.Error()))
+		}
+	} else {
+		// Fully re-bonding the lock
+		// This branch implies that we need to completely replace the old lock in state with a new rebonded lock
+
+		// Restart lock timer and set back to the store
+		// Rebonded lock is the same lock as the original lock, but with an empty EndTime.
+		rebondedLock = types.NewPeriodLock(lock.ID, owner, owner.String(), lock.Duration, time.Time{}, lock.Coins)
+		err = k.setLock(ctx, rebondedLock)
+		if err != nil {
+			return err
+		}
```

Comment:
We should already have the old lock in state, so we should be able to pull the owner and reward address straight from the lock itself. Otherwise this would overwrite the rewards address. Same goes for the previous logic branch for partial relock, but once again I don't think the partial relock is worth the added complexity.